### DISCUSSION TITLE
MXNET on Windows?

### DISCUSSION THREAD

**Discussion by profK** [apache/incubator-mxnet#21131] [General]
The front page says Releases has a pre-built package for windows but all I see there are src packages. Is there still a pre-built package for Windows?

If not, are there proven instructions for building on Windows with VS2022? I have gotten CMAKE to run without errors but building with VS errors out with
an inability to find mxnet-cpp/op.h

Using msbuild I get a long list of link errors:
msbuild -property:USE_CPP_PACKAGE=1 /m:1 ALL_BUILD.vcxproj

I’ve been workign on this for the better part of a week with no success :frowning:

Link errors below:

`    Creating library D:/libraries/incubator-mxnet/build/Debug/mxnet_86.lib and object D:/libraries/incubator-mxnet/bui
  ld/Debug/mxnet_86.exp
crop.cc.obj : error LNK2019: unresolved external symbol "void __cdecl mxnet::op::image::ResizeImplCUDA<class mshadow::b
float::bf16_t,struct mshadow::Tensor<struct mshadow::gpu,3,class mshadow::bfloat::bf16_t>,float>(struct mshadow::Stream
<struct mshadow::gpu> *,struct mshadow::Tensor<struct mshadow::gpu,3,class mshadow::bfloat::bf16_t>,struct mshadow::Ten
sor<struct mshadow::gpu,3,class mshadow::bfloat::bf16_t>)" (??$ResizeImplCUDA@Vbf16_t@bfloat@mshadow@@U?$Tensor@Ugpu@ms
hadow@@$02Vbf16_t@bfloat@2@@3@M@image@op@mxnet@@YAXPEAU?$Stream@Ugpu@mshadow@@@mshadow@@U?$Tensor@Ugpu@mshadow@@$02Vbf1
6_t@bfloat@2@@4@1@Z) referenced in function "void __cdecl mxnet::op::image::ResizeImplWrapper<struct mshadow::cpu>(stru
ct mxnet::op::image::ResizeParam const &,struct mxnet::OpContext const &,class std::vector<class mxnet::TBlob,class std
::allocator<class mxnet::TBlob> > const &,class std::vector<class mxnet::TBlob,class std::allocator<class mxnet::TBlob>
 > const &)" (??$ResizeImplWrapper@Ucpu@mshadow@@@image@op@mxnet@@YAXAEBUResizeParam@012@AEBUOpContext@2@AEBV?$vector@V
TBlob@mxnet@@V?$allocator@VTBlob@mxnet@@@std@@@std@@2@Z) [D:\libraries\incubator-mxnet\build\mxnet_86.vcxproj]
resize.cc.obj : error LNK2001: unresolved external symbol "void __cdecl mxnet::op::image::ResizeImplCUDA<class mshadow:
:bfloat::bf16_t,struct mshadow::Tensor<struct mshadow::gpu,3,class mshadow::bfloat::bf16_t>,float>(struct mshadow::Stre
am<struct mshadow::gpu> *,struct mshadow::Tensor<struct mshadow::gpu,3,class mshadow::bfloat::bf16_t>,struct mshadow::T
ensor<struct mshadow::gpu,3,class mshadow::bfloat::bf16_t>)" (??$ResizeImplCUDA@Vbf16_t@bfloat@mshadow@@U?$Tensor@Ugpu@
mshadow@@$02Vbf16_t@bfloat@2@@3@M@image@op@mxnet@@YAXPEAU?$Stream@Ugpu@mshadow@@@mshadow@@U?$Tensor@Ugpu@mshadow@@$02Vb
f16_t@bfloat@2@@4@1@Z) [D:\libraries\incubator-mxnet\build\mxnet_86.vcxproj]
crop.cc.obj : error LNK2019: unresolved external symbol "void __cdecl mxnet::op::image::ResizeImplCUDA<class mshadow::b
float::bf16_t,struct mshadow::Tensor<struct mshadow::gpu,4,class mshadow::bfloat::bf16_t>,float>(struct mshadow::Stream
<struct mshadow::gpu> *,struct mshadow::Tensor<struct mshadow::gpu,4,class mshadow::bfloat::bf16_t>,struct mshadow::Ten
sor<struct mshadow::gpu,4,class mshadow::bfloat::bf16_t>)" (??$ResizeImplCUDA@Vbf16_t@bfloat@mshadow@@U?$Tensor@Ugpu@ms
hadow@@$03Vbf16_t@bfloat@2@@3@M@image@op@mxnet@@YAXPEAU?$Stream@Ugpu@mshadow@@@mshadow@@U?$Tensor@Ugpu@mshadow@@$03Vbf1
6_t@bfloat@2@@4@1@Z) referenced in function "void __cdecl mxnet::op::image::ResizeImplWrapper<struct mshadow::cpu>(stru
ct mxnet::op::image::ResizeParam const &,struct mxnet::OpContext const &,class std::vector<class mxnet::TBlob,class std
::allocator<class mxnet::TBlob> > const &,class std::vector<class mxnet::TBlob,class std::allocator<class mxnet::TBlob>
 > const &)" (??$ResizeImplWrapper@Ucpu@mshadow@@@image@op@mxnet@@YAXAEBUResizeParam@012@AEBUOpContext@2@AEBV?$vector@V
TBlob@mxnet@@V?$allocator@VTBlob@mxnet@@@std@@@std@@2@Z) [D:\libraries\incubator-mxnet\build\mxnet_86.vcxproj]
resize.cc.obj : error LNK2001: unresolved external symbol "void __cdecl mxnet::op::image::ResizeImplCUDA<class mshadow:
:bfloat::bf16_t,struct mshadow::Tensor<struct mshadow::gpu,4,class mshadow::bfloat::bf16_t>,float>(struct mshadow::Stre
am<struct mshadow::gpu> *,struct mshadow::Tensor<struct mshadow::gpu,4,class mshadow::bfloat::bf16_t>,struct mshadow::T
ensor<struct mshadow::gpu,4,class mshadow::bfloat::bf16_t>)" (??$ResizeImplCUDA@Vbf16_t@bfloat@mshadow@@U?$Tensor@Ugpu@
mshadow@@$03Vbf16_t@bfloat@2@@3@M@image@op@mxnet@@YAXPEAU?$Stream@Ugpu@mshadow@@@mshadow@@U?$Tensor@Ugpu@mshadow@@$03Vb
f16_t@bfloat@2@@4@1@Z) [D:\libraries\incubator-mxnet\build\mxnet_86.vcxproj]
image_random.cc.obj : error LNK2019: unresolved external symbol "void __cdecl mxnet::op::image::ToTensorImplCUDA<class
mshadow::bfloat::bf16_t,struct mshadow::Tensor<struct mshadow::gpu,3,class mshadow::bfloat::bf16_t>,struct mshadow::Ten
sor<struct mshadow::gpu,3,float> >(struct mshadow::Stream<struct mshadow::gpu> *,struct mshadow::Tensor<struct mshadow:
:gpu,3,class mshadow::bfloat::bf16_t>,struct mshadow::Tensor<struct mshadow::gpu,3,float>,int,float)" (??$ToTensorImplC
UDA@Vbf16_t@bfloat@mshadow@@U?$Tensor@Ugpu@mshadow@@$02Vbf16_t@bfloat@2@@3@U?$Tensor@Ugpu@mshadow@@$02M@3@@image@op@mxn
et@@YAXPEAU?$Stream@Ugpu@mshadow@@@mshadow@@U?$Tensor@Ugpu@mshadow@@$02Vbf16_t@bfloat@2@@4@U?$Tensor@Ugpu@mshadow@@$02M
@4@HM@Z) referenced in function "void __cdecl mxnet::op::image::ToTensorOpForward<struct mshadow::cpu>(struct nnvm::Nod
eAttrs const &,struct mxnet::OpContext const &,class std::vector<class mxnet::TBlob,class std::allocator<class mxnet::T
Blob> > const &,class std::vector<enum mxnet::OpReqType,class std::allocator<enum mxnet::OpReqType> > const &,class std
::vector<class mxnet::TBlob,class std::allocator<class mxnet::TBlob> > const &)" (??$ToTensorOpForward@Ucpu@mshadow@@@i
mage@op@mxnet@@YAXAEBUNodeAttrs@nnvm@@AEBUOpContext@2@AEBV?$vector@VTBlob@mxnet@@V?$allocator@VTBlob@mxnet@@@std@@@std@
@AEBV?$vector@W4OpReqType@mxnet@@V?$allocator@W4OpReqType@mxnet@@@std@@@7@2@Z) [D:\libraries\incubator-mxnet\build\mxne
t_86.vcxproj]
image_random.cc.obj : error LNK2019: unresolved external symbol "void __cdecl mxnet::op::image::ToTensorImplCUDA<class
mshadow::bfloat::bf16_t,struct mshadow::Tensor<struct mshadow::gpu,4,class mshadow::bfloat::bf16_t>,struct mshadow::Ten
sor<struct mshadow::gpu,4,float> >(struct mshadow::Stream<struct mshadow::gpu> *,struct mshadow::Tensor<struct mshadow:
:gpu,4,class mshadow::bfloat::bf16_t>,struct mshadow::Tensor<struct mshadow::gpu,4,float>,int,float)" (??$ToTensorImplC
UDA@Vbf16_t@bfloat@mshadow@@U?$Tensor@Ugpu@mshadow@@$03Vbf16_t@bfloat@2@@3@U?$Tensor@Ugpu@mshadow@@$03M@3@@image@op@mxn
et@@YAXPEAU?$Stream@Ugpu@mshadow@@@mshadow@@U?$Tensor@Ugpu@mshadow@@$03Vbf16_t@bfloat@2@@4@U?$Tensor@Ugpu@mshadow@@$03M
@4@HM@Z) referenced in function "void __cdecl mxnet::op::image::ToTensorOpForward<struct mshadow::cpu>(struct nnvm::Nod
eAttrs const &,struct mxnet::OpContext const &,class std::vector<class mxnet::TBlob,class std::allocator<class mxnet::T
Blob> > const &,class std::vector<enum mxnet::OpReqType,class std::allocator<enum mxnet::OpReqType> > const &,class std
::vector<class mxnet::TBlob,class std::allocator<class mxnet::TBlob> > const &)" (??$ToTensorOpForward@Ucpu@mshadow@@@i
mage@op@mxnet@@YAXAEBUNodeAttrs@nnvm@@AEBUOpContext@2@AEBV?$vector@VTBlob@mxnet@@V?$allocator@VTBlob@mxnet@@@std@@@std@
@AEBV?$vector@W4OpReqType@mxnet@@V?$allocator@W4OpReqType@mxnet@@@std@@@7@2@Z) [D:\libraries\incubator-mxnet\build\mxne
t_86.vcxproj]
image_random.cc.obj : error LNK2019: unresolved external symbol "void __cdecl mxnet::op::image::NormalizeImplCUDA<class
 mshadow::bfloat::bf16_t>(struct mshadow::Stream<struct mshadow::gpu> *,class mshadow::bfloat::bf16_t const *,class msh
adow::bfloat::bf16_t *,int,int,int,int,int,float,float,float,float,float,float)" (??$NormalizeImplCUDA@Vbf16_t@bfloat@m
shadow@@@image@op@mxnet@@YAXPEAU?$Stream@Ugpu@mshadow@@@mshadow@@PEBVbf16_t@bfloat@4@PEAV564@HHHHHMMMMMM@Z) referenced
in function "void __cdecl mxnet::op::image::NormalizeOpForward<struct mshadow::cpu>(struct nnvm::NodeAttrs const &,stru
ct mxnet::OpContext const &,class std::vector<class mxnet::TBlob,class std::allocator<class mxnet::TBlob> > const &,cla
ss std::vector<enum mxnet::OpReqType,class std::allocator<enum mxnet::OpReqType> > const &,class std::vector<class mxne
t::TBlob,class std::allocator<class mxnet::TBlob> > const &)" (??$NormalizeOpForward@Ucpu@mshadow@@@image@op@mxnet@@YAX
AEBUNodeAttrs@nnvm@@AEBUOpContext@2@AEBV?$vector@VTBlob@mxnet@@V?$allocator@VTBlob@mxnet@@@std@@@std@@AEBV?$vector@W4Op
ReqType@mxnet@@V?$allocator@W4OpReqType@mxnet@@@std@@@7@2@Z) [D:\libraries\incubator-mxnet\build\mxnet_86.vcxproj]
image_random.cc.obj : error LNK2019: unresolved external symbol "void __cdecl mxnet::op::image::NormalizeBackwardImplCU
DA<class mshadow::bfloat::bf16_t>(struct mshadow::Stream<struct mshadow::gpu> *,class mshadow::bfloat::bf16_t const *,c
lass mshadow::bfloat::bf16_t *,int,int,int,int,int,float,float,float)" (??$NormalizeBackwardImplCUDA@Vbf16_t@bfloat@msh
adow@@@image@op@mxnet@@YAXPEAU?$Stream@Ugpu@mshadow@@@mshadow@@PEBVbf16_t@bfloat@4@PEAV564@HHHHHMMM@Z) referenced in fu
nction "void __cdecl mxnet::op::image::NormalizeOpBackward<struct mshadow::cpu>(struct nnvm::NodeAttrs const &,struct m
xnet::OpContext const &,class std::vector<class mxnet::TBlob,class std::allocator<class mxnet::TBlob> > const &,class s
td::vector<enum mxnet::OpReqType,class std::allocator<enum mxnet::OpReqType> > const &,class std::vector<class mxnet::T
Blob,class std::allocator<class mxnet::TBlob> > const &)" (??$NormalizeOpBackward@Ucpu@mshadow@@@image@op@mxnet@@YAXAEB
UNodeAttrs@nnvm@@AEBUOpContext@2@AEBV?$vector@VTBlob@mxnet@@V?$allocator@VTBlob@mxnet@@@std@@@std@@AEBV?$vector@W4OpReq
Type@mxnet@@V?$allocator@W4OpReqType@mxnet@@@std@@@7@2@Z) [D:\libraries\incubator-mxnet\build\mxnet_86.vcxproj]
np_multinomial_op.cc.obj : error LNK2019: unresolved external symbol "void __cdecl mxnet::op::CheckPvalGPU<class mshado
w::bfloat::bf16_t>(struct mxnet::OpContext const &,class mshadow::bfloat::bf16_t *,int)" (??$CheckPvalGPU@Vbf16_t@bfloa
t@mshadow@@@op@mxnet@@YAXAEBUOpContext@1@PEAVbf16_t@bfloat@mshadow@@H@Z) referenced in function "void __cdecl mxnet::op
::NumpyMultinomialForward<struct mshadow::cpu>(struct nnvm::NodeAttrs const &,struct mxnet::OpContext const &,class std
::vector<class mxnet::TBlob,class std::allocator<class mxnet::TBlob> > const &,class std::vector<enum mxnet::OpReqType,
class std::allocator<enum mxnet::OpReqType> > const &,class std::vector<class mxnet::TBlob,class std::allocator<class m
xnet::TBlob> > const &)" (??$NumpyMultinomialForward@Ucpu@mshadow@@@op@mxnet@@YAXAEBUNodeAttrs@nnvm@@AEBUOpContext@1@AE
BV?$vector@VTBlob@mxnet@@V?$allocator@VTBlob@mxnet@@@std@@@std@@AEBV?$vector@W4OpReqType@mxnet@@V?$allocator@W4OpReqTyp
e@mxnet@@@std@@@6@2@Z) [D:\libraries\incubator-mxnet\build\mxnet_86.vcxproj]
D:\libraries\incubator-mxnet\build\Debug\mxnet_86.dll : fatal error LNK1120: 7 unresolved externals [D:\libraries\incub
ator-mxnet\build\mxnet_86.vcxproj]`

**Comment by profK** [comment]
Update.  I figured out that the script was looking for python3.exe whereas  my python 3 is just called python.exe so I copied it to python3.exe.  Rebuilding, fingers crossed.

So, now I am gtting this...

`
Running: OpWrapperGenerator.py"
24>Traceback (most recent call last):

24>  File "D:\libraries\incubator-mxnet\cpp-package\scripts\OpWrapperGenerator.py", line 434, in <module>

24>    raise(e)

24>  File "D:\libraries\incubator-mxnet\cpp-package\scripts\OpWrapperGenerator.py", line 428, in <module>

24>    f.write(patternStr % ParseAllOps())

24>  File "D:\libraries\incubator-mxnet\cpp-package\scripts\OpWrapperGenerator.py", line 323, in ParseAllOps

24>    cdll.libmxnet = cdll.LoadLibrary(sys.argv[1])

24>  File "C:\Users\jpkessel\AppData\Local\Programs\Python\Python39\lib\ctypes\__init__.py", line 452, in LoadLibrary

24>    return self._dlltype(name)

24>  File "C:\Users\jpkessel\AppData\Local\Programs\Python\Python39\lib\ctypes\__init__.py", line 374, in __init__

24>    self._handle = _dlopen(self._name, mode)

24>FileNotFoundError: Could not find module 'D:\libraries\incubator-mxnet\build\Release\libmxnet.dll' (or one of its dependencies). Try using the full path with constru
`

**Reply by profK** [reply]
Update: looks like its one of the dependencies because it is there...

**Comment by relaxli00** [comment]
OpWrapperGenerator.py needs load libmxnet.dll. It will need all the dependencies either under the same folder or under the environment path.

I manually build 1.9.1 with MKL BLAS using VS2019 recently. The steps and dependencies I've figured out can be found at https://medium.com/@relaxli/build-mxnet-1-9-1-with-cuda-11-7-and-mkl-blas-using-visual-studio-2019-on-windows-10-a9ebecb1941b

**Comment by profK** [comment]
Thanks.  In the end I went a different route for this project, but I
might come back to mxnet at some point

On Wed, Aug 31, 2022 at 7:50 PM relaxli00 ***@***.***> wrote:

> OpWrapperGenerator.py needs load libmxnet.dll. It will need all the
> dependencies either under the same folder or under the environment path.
>
> I manually build 1.9.1 with MKL BLAS using VS2019 recently. The steps and
> dependencies I've figured out can be found at
> ***@***.***/build-mxnet-1-9-1-with-cuda-11-7-and-mkl-blas-using-visual-studio-2019-on-windows-10-a9ebecb1941b
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-mxnet/discussions/21131#discussioncomment-3522321>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA64NZKXDKQJN7ZTCTTFH2TV37VULANCNFSM57QD6WIA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>


-- 
It's always darkest just before you are eaten by a grue.
